docs/docs/preprocessing-external-images.md

```diff
@@ -0,0 +1,170 @@
+  if (
+    node.internal.type === "MarkdownRemark" &&
+    node.frontmatter.featuredImgUrl !== null
+  ) {
+    let fileNode = await createRemoteFileNode({
+      url: node.frontmatter.featuredImgUrl, // string that points to the URL of the image
+      parentNodeId: node.id, // id of the parent node of the fileNode we are going to create
+      createNode, // helper function in gatsby-node to generate the node
+      createNodeId, // helper function in gatsby-node to generate the node id
+      cache, // Gatsby's cache
+      store, // Gatsby's redux store
+    })
+
+    // if the file was created, attach the new node to the parent node
+    if (fileNode) {
+      node.featuredImg___NODE = fileNode.id
+    }
+  }
+}
+```
+
+Going step by step through the code:
+
+1. Define some types for `MarkdownRemark` using the [Schema Customization API](/docs/schema-customization/) so if `featuredImgUrl` is not in a markdown file, it will return null.
+2. Create an `onCreateNode` function so you can watch for when `MarkdownRemark` nodes are made.
+3. Use `createRemoteFileNode` by passing in the various required fields and get a reference to it afterwards.
+4. If the node is created, attach it as a child of the original node. `___NODE` tells the graphql layer that the name before it is going to be a field on the parent node that links to another node. To do this, pass the id as the reference.
+
+And now since it is a file node, `gatsby-transformer-sharp` will pick it up and create a `childImageSharp` child node inside this newly created node.
+
+## Usage in templates
+
+Now that the images are being generated and available in GraphQL, let's use it in action.
+
+If you open GraphiQL and write a query on the markdown nodes, you can see a new node attached to any `MarkdownRemark` node that had a featured image:
+
+```graphql
+query {
+  allMarkdownRemark {
+    nodes {
+      featuredImg {
+        childImageSharp {
+          # ...
+        }
+      }
+    }
+  }
+}
+```
+
+![Screenshot of GraphiQL with above query inserted](images/remote-file-node-graphiql-preview.png)
+
+We can then use `gatsby-transformer-sharp` to fill in the query for a fixed image here.
+
+```graphql
+query {
+  allMarkdownRemark {
+    nodes {
+      featuredImg {
+        childImageSharp {
+          fixed(width: 600) {
+            ...GatsbyImageSharpFixed
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+And finally, we can update the template for this blogpost to include this image. This template is based on the one in the [Programmatically create pages from data](/tutorial/part-seven/) section of the Gatsby Tutorial.
+
+```jsx
+import React from "react"
+import Img from "gatsby-image"
+import { graphql } from "gatsby"
+
+const template = ({ data }) => {
+  return (
+    <>
+      <h1>{data.markdownRemark.frontmatter.title}</h1>
+      <Img fixed={data.markdownRemark.featuredImg.childImageSharp.fixed} />
```

Comment:
It would be a good opportunity to talk about sourcing alt text from Frontmatter for these images. Can that be factored into the new Node type?

Reply:
Yeah definitely!

Reply:
And done!